engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala

```
@@ -0,0 +1,194 @@
+  ModuleReference
+}
+
+import scala.collection.immutable.ListMap
+import scala.collection.mutable
+
+/** The module allowing to resolve the dependencies between the component groups
+  * of different packages.
+  */
+final class ComponentGroupsResolver {
+
+  /** Run the component groups resolution algorithm.
+    *
+    * A package can define a new component group or extend an existing one. The
+    * resolving algorithm takes the component groups defined by the packages and
+    * applies the available extensions.
+    *
+    * @param packages the list of package configs
+    * @return the list of component groups with the dependencies resolved
+    */
+  def run(packages: Iterable[Config]): Vector[LibraryComponentGroup] = {
+    val libraryComponents = packages
+      .map { config =>
+        LibraryName(config.namespace, config.name) -> config.componentGroups
+      }
+      .collect { case (libraryName, Right(componentGroups)) =>
+        libraryName -> componentGroups
+      }
+    val libraryComponentsMap =
+      ComponentGroupsResolver
+        .toMapKeepFirst(libraryComponents)
+        .to(ListMap)
+    resolveComponentGroups(libraryComponentsMap)
+  }
+
+  /** Resolve the component groups. Utility method that takes a list of
+    * component groups associated with the library name.
+    *
+    * @param libraryComponents the associated list of component groups
+    * @return the list of component groups with dependencies resolved
+    */
+  private def resolveComponentGroups(
+    libraryComponents: Map[LibraryName, ComponentGroups]
+  ): Vector[LibraryComponentGroup] = {
+    val newLibraryComponentGroups =
+      libraryComponents.view
+        .flatMap { case (libraryName, componentGroups) =>
+          componentGroups.newGroups.map(toLibraryComponentGroup(libraryName, _))
+        }
+    val newLibraryComponentGroupsMap =
+      ComponentGroupsResolver
+        .groupByKeepFirst(newLibraryComponentGroups) { libraryComponentGroup =>
```

Again - is it ever possible that the `ModuleReference` will be duplicated?
We already deduplicated the libraries earlier, so at this stage a library would need to declare duplicated component groups, right? If this is even allowed, shouldn't a warning be reported somewhere?

It feels like silently ignoring entries like these may lead to hard to detect errors somewhere down the line (but maybe in this particular case I'm misunderstanding something?)